feflow/utils/hybrid_topology.py

```diff
@@ -56,23 +124,2437 @@ def __init__(self,
+                const_idx)
+            hybrid_atoms = tuple(sorted([hybrid_map[at1], hybrid_map[at2]]))
+            if hybrid_atoms not in constraint_lengths.keys():
+                self._hybrid_system.addConstraint(hybrid_atoms[0],
+                                                  hybrid_atoms[1], length)
+                constraint_lengths[hybrid_atoms] = length
+            else:
+                if constraint_lengths[hybrid_atoms] != length:
+                    raise AssertionError('constraint length is changing')
+
+    @staticmethod
+    def _copy_threeparticleavg(atm_map, env_atoms, vs):
+        """
+        Helper method to copy a ThreeParticleAverageSite virtual site
+        from two mapped Systems.
+
+        Parameters
+        ----------
+        atm_map : dict[int, int]
+          The atom map correspondance between the two Systems.
+        env_atoms: set[int]
+          A list of environment atoms for the target System. This
+          checks that no alchemical atoms are being tied to.
+        vs : openmm.ThreeParticleAverageSite
+
+        Returns
+        -------
+        openmm.ThreeParticleAverageSite
+        """
+        particles = {}
+        weights = {}
+        for i in range(vs.getNumParticles()):
+            particles[i] = atm_map[vs.getParticle(i)]
+            weights[i] = vs.getWeight(i)
+        if not all(i in env_atoms for i in particles.values()):
+            errmsg = ("Virtual sites bound to non-environment atoms "
+                      "are not supported")
+            raise ValueError(errmsg)
+        return openmm.ThreeParticleAverageSite(
+            particles[0], particles[1], particles[2],
+            weights[0], weights[1], weights[2],
+        )
+
+    def _handle_virtual_sites(self):
+        """
+        Ensure that all virtual sites in old and new system are copied over to
+        the hybrid system. Note that we do not support virtual sites in the
+        changing region.
+
+        TODO - remerge into a single loop
+        TODO - check that it's fine to double count here (even so, there's
+               an optimisation that could be done here...)
+        """
+        # old system
+        # Loop through virtual sites
+        for particle_idx in range(self._old_system.getNumParticles()):
+            if self._old_system.isVirtualSite(particle_idx):
+                # If it's a virtual site, make sure it is not in the unique or
+                # core atoms, since this is currently unsupported
+                hybrid_idx = self._old_to_hybrid_map[particle_idx]
+                if hybrid_idx not in self._atom_classes['environment_atoms']:
+                    errmsg = ("Virtual sites in changing residue are "
+                              "unsupported.")
+                    raise ValueError(errmsg)
+                else:
+                    virtual_site = self._old_system.getVirtualSite(
+                        particle_idx)
```

Comment:
Just a random thought - if you like autoformatters @ijpulidos, now might be a good time to pass this file through `black` or similar.

Reply:
This is whatever my IDE uses (I think it's not `black`), but you are right, maybe we should set up a pre-commit hooks bot here for formatting the code. Hopefully one that doesn't get in the way so much, I'll try setting that up.